### PR TITLE
#162162963 create password reset functionality

### PR DESCRIPTION
**Pull request by aayebare**
### What does this PR do?
- Allows users to reset their password via email
### Description of Task to be completed?
- create password reset form
- create email sending form
- send a link to the user's email to redirect to the password reset form
### How should this be manually tested?
- Clone this repository and run ```yarn install``` so as to install all dependencies
- Run ```yarn start``` to run the server. The default page is the home page at ```/```
- Use ```localhost/reset-password``` to view the email sharing form
- A user is sent a link and on clicking the link, they're redirected to the  ```localhost/reset-password/:token``` to view the password reset form

### snapshots
image

image

### What are the relevant pivotal tracker stories?
#162162963

### CODE REVIEW

File: src/components/auth/index.jsx [hunk 1]
```diff
@@ -10,10 +10,11 @@ const SignUpForm = ({ onSubmit, isSignUp, handleLogin, renderInput }) => (
         </div>
         <div className="card-body">
           <h4 className="sign-up" align="center">
-            {isSignUp ? "Sign Up" : "Log In"}
+          {title}
+            {/* {isSignUp ? "Sign Up" : "Log In"} */}
```
Comment (matthewacha) [inline src/components/auth/index.jsx]: remove commented out code

File: src/redux/reducers/authentication/__tests__/authReducer.test.jsx [hunk 2]
```diff
@@ -47,6 +47,7 @@ describe('Authentication reducers', () => {
       signUpSuccess: null,
     };
     const newState = authReducer(initialState, action);
+    console.log(newState)
```
Comment (matthewacha) [inline src/redux/reducers/authentication/__tests__/authReducer.test.jsx]: remove this

File: src/redux/thunks/index.js [hunk 3]
```diff
@@ -5,24 +5,72 @@ const axiosInstance = axios.create({
   baseURL: "https://ah-backend-poseidon-staging.herokuapp.com/api/"
 });
 
+// axiosInstance.interceptors.response.use(
```
Comment (matthewacha) [inline src/redux/thunks/index.js]: delete these commented out code

File: src/redux/thunks/index.js [hunk 4]
```diff
@@ -5,24 +5,72 @@ const axiosInstance = axios.create({
...
 
-export default postDataThunkNoHeader;
+  
+    const getDataThunk = (endpoint, actionCreator) => {
+      return (dispatch) =>{
+          axiosInstance.get(endpoint).then((response) => {
+              dispatch(actionCreator(response.data));
+          }).catch((error) =>{
+            //   console.log(err);
+            //  dispatch(errorOccurred(err));
```
Comment (matthewacha) [inline src/redux/thunks/index.js]: remove these if they are irrelevant

File: src/views/PasswordResetView.jsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        const { signUpData, signUpErrors } = nextProps;
+        signUpData
+        ? this.redirectOnSuccesfullPasswordReset(signUpData)
+        : notify.show(signUpErrors.detail, 'error', 4000);
+        this.setState({ loader: { loading: false } } );
+}
+
+    componentDidMount=()=>{
+        if (window.location.href.includes('password-reset')) {
+            this.setState({addNewPassword: false});
```
Comment (matthewacha) [inline src/views/PasswordResetView.jsx]: format this properly

File: src/views/PasswordResetView.jsx [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+                title={title}
+                required={required}
+                autoFocus={autoFocus}
+              />
+            </div>
+          );
+      
+       render() {
+        const { isResetPassword} = this.state;
+           const buttonName =  !isResetPassword? "Send Link" : "Reset Password"
```
Comment (matthewacha) [inline src/views/PasswordResetView.jsx]: add a space before '?'

File: src/views/PasswordResetView.jsx [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+               <Fragment>
+               <CircularProgressLoader
+                 {...loader}
+               />
+               <SignUpForm
+               onChange={this.onChange}
+               onSubmit={handleSubmit}
+               isResetPassword={this.state.isResetPassword}
+               changeToResetPassword={this.changeToResetPassword}
+               addNewPassword={this.state.addNewPassword}
```
Comment (matthewacha) [inline src/views/PasswordResetView.jsx]: destructure addNewPassword, isResetPassword, and isSignUp from the state

File: src/views/__tests__/passwordReset.test.jsx [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        Object.defineProperty(window, 'location', {
+        writable:true,
+        value: { href:'http://localhost/password-reset' }         
+        });     
+
+        wrapper.instance().componentDidMount(wrapper.setState({
+            isResetPassword: false,
+            addNewPassword: true
+        })
+)
```
Comment (matthewacha) [inline src/views/__tests__/passwordReset.test.jsx]: what is this test checking for exactly

File: src/views/__tests__/passwordReset.test.jsx [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        
+      });
+   
+    it("tests reset of password link sent to email", () => {
+        wrapper.setState({
+            isResetPassword: false,
+            addNewPassword: true 
+        })
+            const event = {preventDefault: jest.fn(), target:{elements:{email:{value: "m@gmail.com"}} }}
```
Comment (matthewacha) [inline src/views/__tests__/passwordReset.test.jsx]: format this well

File: src/views/authView.jsx [hunk 10]
```diff
@@ -149,16 +155,23 @@ class AuthView extends React.Component {
...
+    const title = isSignUp ? "Sign Up" : "Log In"
+    const buttonName = isSignUp ? "Sign Up" : "Log In"
     return (
       <Fragment>
         <CircularProgressLoader {...loader} />
         <SignUpForm
           onSubmit={onSubmit}
           {...this.state}
           isSignUp={isSignUp}
+          isResetPassword={this.state.isResetPassword}
```
Comment (matthewacha) [inline src/views/authView.jsx:167]: destructure please

**Comment by katunold** [comment]
well done Mark However, I think you need to add screenshots to your PR message showing the steps to reset the password.

**Comment by Luleherll** [comment]
Good work. There some logs on travis that I think you should fix if you can.
screen shot 2019-01-23 at 9 39 16 am

**Review by Luleherll — APPROVED** [review]

**Review by katunold — APPROVED** [review]

**Review by arthurarty — APPROVED** [review]

**Review by billkabanga — APPROVED** [review]

**Review by matthewacha — APPROVED** [review]

**Comment by matthewacha** [comment]
Resolve conflicts

File: src/views/PasswordResetView.jsx [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+          const { loader } = this.state;
+          return (
+            <Fragment>
+              <CircularProgressLoader
+                {...loader}
+              />
+              <SignUpForm
+                onChange={this.onChange}
+                onSubmit={handleSubmit}
+                isResetPassword={this.state.isResetPassword}
```
Comment (matthewacha) [inline src/views/PasswordResetView.jsx]: destructure these, anything with `this.state`, `this.props`

**Pull request merged**